 ### Porto Seguro’s Safe Driver Prediction 

A lot of real business data lies in structured format and the use of deeplearning is not cool (does not gets press coverage) in the community. This notbook will try and reproduce 1st place solution to [Porto Seguro’s Safe Driver Prediction](https://www.kaggle.com/c/porto-seguro-safe-driver-prediction/data) Kaggle competition. 
The overview of the solution is [here](https://www.kaggle.com/c/porto-seguro-safe-driver-prediction/discussion/44629) 

In [1]:
import pandas as pd 
from pandas_summary import DataFrameSummary 

In [2]:
train = pd.read_csv('./train.csv')

In [3]:
train.head(10)

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0   7       0          2              2          5              1   
1   9       0          1              1          7              0   
2  13       0          5              4          9              1   
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   
5  19       0          5              1          4              0   
6  20       0          2              1          3              1   
7  22       0          5              1          4              0   
8  26       0          5              1          3              1   
9  28       1          1              1          2              0   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin       ...        \
0              0              0              1              0       ...         
1              0              0              0              1       ...         
2              0              0              0              1       ...         
3              0              1              0              0       ...         
4              0              1              0              0       ...         
5              0              0              0              0       ...         
6              0              0              1              0       ...         
7              0              1              0              0       ...         
8              0              0              0              1       ...         
9              0              0              1              0       ...         

   ps_calc_11  ps_calc_12  ps_calc_13  ps_calc_14  ps_calc_15_bin  \
0           9           1           5           8               0   
1           3           1           1           9               0   
2           4           2           7           7               0   
3           2           2           4           9               0   
4           3           1           1           3               0   
5           4           2           0           9               0   
6           3           0           0          10               0   
7           7           1           3           6               1   
8           4           2           1           5               0   
9           3           5           0           6               0   

   ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  ps_calc_19_bin  \
0               1               1               0               0   
1               1               1               0               1   
2               1               1               0               1   
3               0               0               0               0   
4               0               0               1               1   
5               1               0               1               1   
6               1               0               0               1   
7               0               1               0               1   
8               1               0               0               0   
9               1               0               0               1   

   ps_calc_20_bin  
0               1  
1               0  
2               0  
3               0  
4               0  
5               1  
6               0  
7               0  
8               1  
9               0  

[10 rows x 59 columns]

In [4]:
train.columns

Index(['id', 'target', 'ps_ind_01', 'ps_ind_02_cat', 'ps_ind_03',
       'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_ind_06_bin', 'ps_ind_07_bin',
       'ps_ind_08_bin', 'ps_ind_09_bin', 'ps_ind_10_bin', 'ps_ind_11_bin',
       'ps_ind_12_bin', 'ps_ind_13_bin', 'ps_ind_14', 'ps_ind_15',
       'ps_ind_16_bin', 'ps_ind_17_bin', 'ps_ind_18_bin', 'ps_reg_01',
       'ps_reg_02', 'ps_reg_03', 'ps_car_01_cat', 'ps_car_02_cat',
       'ps_car_03_cat', 'ps_car_04_cat', 'ps_car_05_cat', 'ps_car_06_cat',
       'ps_car_07_cat', 'ps_car_08_cat', 'ps_car_09_cat', 'ps_car_10_cat',
       'ps_car_11_cat', 'ps_car_11', 'ps_car_12', 'ps_car_13', 'ps_car_14',
       'ps_car_15', 'ps_calc_01', 'ps_calc_02', 'ps_calc_03', 'ps_calc_04',
       'ps_calc_05', 'ps_calc_06', 'ps_calc_07', 'ps_calc_08', 'ps_calc_09',
       'ps_calc_10', 'ps_calc_11', 'ps_calc_12', 'ps_calc_13', 'ps_calc_14',
       'ps_calc_15_bin', 'ps_calc_16_bin', 'ps_calc_17_bin', 'ps_calc_18_bin',
       'ps_calc_19_bin', 'ps_calc_20_bin'],


In [5]:
DataFrameSummary(train).summary()

id     target ps_ind_01 ps_ind_02_cat ps_ind_03  \
count              595212     595212    595212        595212    595212   
mean               743804  0.0364475   1.90038       1.35894   4.42332   
std                429368   0.187401   1.98379      0.664594    2.6999   
min                     7          0         0            -1         0   
25%                371992          0         0             1         2   
50%                743548          0         1             1         4   
75%           1.11555e+06          0         3             2         6   
max           1.48803e+06          1         7             4        11   
counts             595212     595212    595212        595212    595212   
uniques            595212          2         8             5        12   
missing                 0          0         0             0         0   
missing_perc           0%         0%        0%            0%        0%   
types             numeric       bool   numeric       numeric   numeric   

             ps_ind_04_cat ps_ind_05_cat ps_ind_06_bin ps_ind_07_bin  \
count               595212        595212        595212        595212   
mean              0.416794      0.405188      0.393742      0.257033   
std               0.493311       1.35064      0.488579      0.436998   
min                     -1            -1             0             0   
25%                      0             0             0             0   
50%                      0             0             0             0   
75%                      1             0             1             1   
max                      1             6             1             1   
counts              595212        595212        595212        595212   
uniques                  3             8             2             2   
missing                  0             0             0             0   
missing_perc            0%            0%            0%            0%   
types              numeric       numeric          bool          bool   

             ps_ind_08_bin      ...       ps_calc_11 ps_calc_12 ps_calc_13  \
count               595212      ...           595212     595212     595212   
mean              0.163921      ...          5.44138    1.44192    2.87229   
std               0.370205      ...          2.33287    1.20296    1.69489   
min                      0      ...                0          0          0   
25%                      0      ...                4          1          2   
50%                      0      ...                5          1          3   
75%                      0      ...                7          2          4   
max                      1      ...               19         10         13   
counts              595212      ...           595212     595212     595212   
uniques                  2      ...               20         11         14   
missing                  0      ...                0          0          0   
missing_perc            0%      ...               0%         0%         0%   
types                 bool      ...          numeric    numeric    numeric   

             ps_calc_14 ps_calc_15_bin ps_calc_16_bin ps_calc_17_bin  \
count            595212         595212         595212         595212   
mean            7.53903       0.122427        0.62784       0.554182   
std             2.74665       0.327779       0.483381       0.497056   
min                   0              0              0              0   
25%                   6              0              0              0   
50%                   7              0              1              1   
75%                   9              0              1              1   
max                  23              1              1              1   
counts           595212         595212         595212         595212   
uniques              24              2              2              2   
missing               0              0              0              0   
missing_perc        

- Here are lots of features here. In total, our training data has 59 variables, including id and target. In some of them we already see a number of NAs. 

- The data description mentions that the names of the features indicate whether they are binary (bin) or categorical (cat) variables. Everything else is continuous or ordinal.

- We have already been told by Adriano Moala that the names of the variables indicate certain properties: *Ind" is related to individual or driver, “reg” is related to region, “car” is related to car itself and “calc” is an calculated feature.’ Here we will refer to these properties as groups.

These names refer to the origin of variable, but there is no reason to concern about it. "Ind" is related to individual or driver, "reg" is related to region, "car" is related to car itself and "calc" is an calculated feature. The features and rows are not time dependent, so "ps_ind_01" and "ps_ind_02" are just labels to hide real names. Unfortunately I can't share the real meanings of variables.

The 1st place solution deletes all the calc features from the remaining features. 

Get all the feature name list 

In [6]:
features = train.columns

In [7]:
len(features)

59

In [8]:
features[3]

'ps_ind_02_cat'

In [9]:
def calculatedFetures(f):
    calc_features = []
    for feature in f:
        if "calc" in feature:
            calc_features.append(feature)
    return calc_features

calc_features = calculatedFetures(features)

In [10]:
def getCoreFeatures(first, second):
    second = set(second)   
    return [item for item in first if item not in second]

core_features = getCoreFeatures(features, calc_features)    

In [11]:
core_features

['id',
 'target',
 'ps_ind_01',
 'ps_ind_02_cat',
 'ps_ind_03',
 'ps_ind_04_cat',
 'ps_ind_05_cat',
 'ps_ind_06_bin',
 'ps_ind_07_bin',
 'ps_ind_08_bin',
 'ps_ind_09_bin',
 'ps_ind_10_bin',
 'ps_ind_11_bin',
 'ps_ind_12_bin',
 'ps_ind_13_bin',
 'ps_ind_14',
 'ps_ind_15',
 'ps_ind_16_bin',
 'ps_ind_17_bin',
 'ps_ind_18_bin',
 'ps_reg_01',
 'ps_reg_02',
 'ps_reg_03',
 'ps_car_01_cat',
 'ps_car_02_cat',
 'ps_car_03_cat',
 'ps_car_04_cat',
 'ps_car_05_cat',
 'ps_car_06_cat',
 'ps_car_07_cat',
 'ps_car_08_cat',
 'ps_car_09_cat',
 'ps_car_10_cat',
 'ps_car_11_cat',
 'ps_car_11',
 'ps_car_12',
 'ps_car_13',
 'ps_car_14',
 'ps_car_15']

In [12]:
len(core_features),len(calc_features) #sanity check that we are missing any other feature

(39, 20)

So we will be working mainly with core_features as mentioned in the solutioning

In [13]:
core_data = train[core_features] ### This is where i get rid of *calc features 

In [14]:
core_data.head()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0   7       0          2              2          5              1   
1   9       0          1              1          7              0   
2  13       0          5              4          9              1   
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin    ...      \
0              0              0              1              0    ...       
1              0              0              0              1    ...       
2              0              0              0              1    ...       
3              0              1              0              0    ...       
4              0              1              0              0    ...       

   ps_car_07_cat  ps_car_08_cat  ps_car_09_cat  ps_car_10_cat  ps_car_11_cat  \
0              1              0              0              1             12   
1              1              1              2              1             19   
2              1              1              2              1             60   
3              1              1              3              1            104   
4              1              1              2              1             82   

   ps_car_11  ps_car_12  ps_car_13  ps_car_14  ps_car_15  
0          2   0.400000   0.883679   0.370810   3.605551  
1          3   0.316228   0.618817   0.388716   2.449490  
2          1   0.316228   0.641586   0.347275   3.316625  
3          1   0.374166   0.542949   0.294958   2.000000  
4          3   0.316070   0.565832   0.365103   2.000000  

[5 rows x 39 columns]

In [15]:
DataFrameSummary(core_data).summary()

id     target ps_ind_01 ps_ind_02_cat ps_ind_03  \
count              595212     595212    595212        595212    595212   
mean               743804  0.0364475   1.90038       1.35894   4.42332   
std                429368   0.187401   1.98379      0.664594    2.6999   
min                     7          0         0            -1         0   
25%                371992          0         0             1         2   
50%                743548          0         1             1         4   
75%           1.11555e+06          0         3             2         6   
max           1.48803e+06          1         7             4        11   
counts             595212     595212    595212        595212    595212   
uniques            595212          2         8             5        12   
missing                 0          0         0             0         0   
missing_perc           0%         0%        0%            0%        0%   
types             numeric       bool   numeric       numeric   numeric   

             ps_ind_04_cat ps_ind_05_cat ps_ind_06_bin ps_ind_07_bin  \
count               595212        595212        595212        595212   
mean              0.416794      0.405188      0.393742      0.257033   
std               0.493311       1.35064      0.488579      0.436998   
min                     -1            -1             0             0   
25%                      0             0             0             0   
50%                      0             0             0             0   
75%                      1             0             1             1   
max                      1             6             1             1   
counts              595212        595212        595212        595212   
uniques                  3             8             2             2   
missing                  0             0             0             0   
missing_perc            0%            0%            0%            0%   
types              numeric       numeric          bool          bool   

             ps_ind_08_bin    ...    ps_car_07_cat ps_car_08_cat  \
count               595212    ...           595212        595212   
mean              0.163921    ...         0.910027       0.83208   
std               0.370205    ...         0.347106      0.373796   
min                      0    ...               -1             0   
25%                      0    ...                1             1   
50%                      0    ...                1             1   
75%                      0    ...                1             1   
max                      1    ...                1             1   
counts              595212    ...           595212        595212   
uniques                  2    ...                3             2   
missing                  0    ...                0             0   
missing_perc            0%    ...               0%            0%   
types                 bool    ...          numeric          bool   

             ps_car_09_cat ps_car_10_cat ps_car_11_cat ps_car_11 ps_car_12  \
count               595212        595212        595212    595212    595212   
mean               1.32889      0.992136       62.2157   2.34607  0.379945   
std               0.978747     0.0916186       33.0125  0.832548  0.058327   
min                     -1             0             1        -1        -1   
25%                      0             1            32         2  0.316228   
50%                      2             1            65         3  0.374166   
75%                      2             1            93         3       0.4   
max                      4             2           104         3   1.26491   
counts              595212        595212        595212    595212    595212   
uniques                  6             3           104         5       184   
missing                  0             0             0         0         0   
missing_perc            0%            0%            0%        0%        0%   
types     

Turning categorical variables and binary variables into one hot encoding 

Get all cat variable

In [16]:
def catFetures(f):
    cat_features = []
    for feature in f:
        if "cat" in feature:
            cat_features.append(feature)
    return cat_features

In [17]:
cat_features = catFetures(core_data.columns)

In [18]:
cat_features

['ps_ind_02_cat',
 'ps_ind_04_cat',
 'ps_ind_05_cat',
 'ps_car_01_cat',
 'ps_car_02_cat',
 'ps_car_03_cat',
 'ps_car_04_cat',
 'ps_car_05_cat',
 'ps_car_06_cat',
 'ps_car_07_cat',
 'ps_car_08_cat',
 'ps_car_09_cat',
 'ps_car_10_cat',
 'ps_car_11_cat']

In [19]:
len(cat_features)

14

In [20]:
def binFetures(f):
    bin_features = []
    for feature in f:
        if "bin" in feature:
            bin_features.append(feature)
    return bin_features

In [21]:
bin_features = binFetures(core_data.columns)

In [22]:
bin_features

['ps_ind_06_bin',
 'ps_ind_07_bin',
 'ps_ind_08_bin',
 'ps_ind_09_bin',
 'ps_ind_10_bin',
 'ps_ind_11_bin',
 'ps_ind_12_bin',
 'ps_ind_13_bin',
 'ps_ind_16_bin',
 'ps_ind_17_bin',
 'ps_ind_18_bin']

In [23]:
len(bin_features)

11

In [24]:
cat_columns = cat_features + bin_features

In [25]:
print(cat_columns)

['ps_ind_02_cat', 'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_car_01_cat', 'ps_car_02_cat', 'ps_car_03_cat', 'ps_car_04_cat', 'ps_car_05_cat', 'ps_car_06_cat', 'ps_car_07_cat', 'ps_car_08_cat', 'ps_car_09_cat', 'ps_car_10_cat', 'ps_car_11_cat', 'ps_ind_06_bin', 'ps_ind_07_bin', 'ps_ind_08_bin', 'ps_ind_09_bin', 'ps_ind_10_bin', 'ps_ind_11_bin', 'ps_ind_12_bin', 'ps_ind_13_bin', 'ps_ind_16_bin', 'ps_ind_17_bin', 'ps_ind_18_bin']


In [26]:
len(cat_columns)

25

In [27]:
def one_hot_encode(core_data,cat_columns):    
    for cat_column in cat_columns: 
        print("Resolving columns", cat_column)
        core_data_with_onehot = pd.get_dummies(core_data[cat_column])
        core_data = pd.concat([core_data, core_data_with_onehot], axis=1)
        del core_data[cat_column]    
    return core_data

core_data = one_hot_encode(core_data, cat_columns)

Resolving columns ps_ind_02_cat
Resolving columns ps_ind_04_cat
Resolving columns ps_ind_05_cat
Resolving columns ps_car_01_cat
Resolving columns ps_car_02_cat
Resolving columns ps_car_03_cat
Resolving columns ps_car_04_cat
Resolving columns ps_car_05_cat
Resolving columns ps_car_06_cat
Resolving columns ps_car_07_cat
Resolving columns ps_car_08_cat
Resolving columns ps_car_09_cat
Resolving columns ps_car_10_cat
Resolving columns ps_car_11_cat
Resolving columns ps_ind_06_bin
Resolving columns ps_ind_07_bin
Resolving columns ps_ind_08_bin
Resolving columns ps_ind_09_bin
Resolving columns ps_ind_10_bin
Resolving columns ps_ind_11_bin
Resolving columns ps_ind_12_bin
Resolving columns ps_ind_13_bin
Resolving columns ps_ind_16_bin
Resolving columns ps_ind_17_bin
Resolving columns ps_ind_18_bin


In [28]:
core_data.head()

id  target  ps_ind_01  ps_ind_03  ps_ind_14  ps_ind_15  ps_reg_01  \
0   7       0          2          5          0         11        0.7   
1   9       0          1          7          0          3        0.8   
2  13       0          5          9          0         12        0.0   
3  16       0          0          2          0          8        0.9   
4  17       0          0          0          0          9        0.7   

   ps_reg_02  ps_reg_03  ps_car_11 ...  0  1  0  1  0  1  0  1  0  1  
0        0.2   0.718070          2 ...  1  0  1  0  1  0  0  1  1  0  
1        0.4   0.766078          3 ...  1  0  1  0  1  0  1  0  0  1  
2        0.0  -1.000000          1 ...  1  0  1  0  0  1  1  0  1  0  
3        0.2   0.580948          1 ...  1  0  1  0  0  1  1  0  1  0  
4        0.6   0.840759          3 ...  1  0  1  0  0  1  1  0  1  0  

[5 rows x 220 columns]

In [29]:
core_data.columns

Index([       'id',    'target', 'ps_ind_01', 'ps_ind_03', 'ps_ind_14',
       'ps_ind_15', 'ps_reg_01', 'ps_reg_02', 'ps_reg_03', 'ps_car_11',
       ...
                 0,           1,           0,           1,           0,
                 1,           0,           1,           0,           1],
      dtype='object', length=220)

We should have 221 features in the data set on which we will train our autoencoder.

 ```This ends up in exactly 221 dense features.```

I am having an error here, I get 220 features.